tests/tests_app/core/test_lightning_app_checkpoint.py

```diff
@@ -0,0 +1,256 @@
+import json
+import os
+from unittest import mock
+
+import pytest
+from tests_app import _PROJECT_ROOT
+
+from lightning.app.structures import Dict
+from lightning_app import LightningApp, LightningFlow, LightningWork  # F401
+
+
+class FlowComponent(LightningFlow):
+    def __init__(self, text):
+        super().__init__()
+        self.text = text
+
+    def run(self):
+        print(f"what is the value of text? {self.text}")
+
+
+class SavedAppExample(LightningFlow):
+    def __init__(self):
+        super().__init__()
+        self.text = ""
+        self.flow_component = FlowComponent("Hello Flow Component!")
+
+    def run(self):
+        print(f"what is the value of text? {self.text}")
+
+
+class DynamicWorkLevel3(LightningWork):
+    def __init__(self, required_arg):
+        super().__init__()
+        self.text = ""
+        self.required_arg = required_arg
+
+    def run(self):
+        print(f"what is the value of text in the work? {self.text}")
+
+
+class DynamicFlowLevel2(LightningFlow):
+    def __init__(self):
+        super().__init__()
+        self.text = ""
+
+    def run(self):
+        if not getattr(self, "work_level_3", None):
+            self.work_level_3 = DynamicWorkLevel3(required_arg="Hello Work Level 3 Required Arg!")
+        self.work_level_3.run()
+
+    def load_state_dict(self, flow_state, children_states, strict) -> None:
+
+        text_from_saved_state = children_states["work_level_3"]["vars"]["text"]
+        assert text_from_saved_state == "Hello Work Level 3!"
+
+        # now we want to update this state with the new value of text and re-instantiate the work with the required args
+        children_states["work_level_3"]["vars"]["text"] = "Hello Work Level 3 Updated!"
```

Comment:
Are we expecting developers implementing `load_state_dict` to manually dig into states this way?

I think we should provide a lightweight API that allows to get var values from the state, otherwise we are coupling `load_state_dict` implementations with the current structure of state. It could be as lightweight as just wrapping the dict in a wrapper class that provides `.` access to "vars" from the state.

I'm good with keeping it this way right now, but I'd open an issue right after merging because we'll need to do something about it in the near future.

Reply:
yes, this will be nice. 
created an issue here: https://github.com/Lightning-AI/lightning/issues/15475